File: app/src/main/java/ir/farshid_roohi/reminderpro/Storage.java
package ir.farshid_roohi.reminderpro;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.json.JSONObject;


/**
 * Created by ImanX.
 * MyZarinPal-Android-V4 | Copyrights 2018 ZarinPal Crop.
 */

public class Storage {

    private SharedPreferences        sp;
    private SharedPreferences.Editor editor;

    public Storage(Context context) {
        this(context, "dev");
    }

    @SuppressLint("CommitPrefEdits")
    public Storage(Context context, String name) {
        this.sp = context.getSharedPreferences(name, Context.MODE_PRIVATE);
        this.editor = sp.edit();
    }

    public <T> void put(Class<T> clazz, String key, @NonNull T value) {
        editor.putString(key, value.toString());
        editor.apply();
    }


    public void remove(String key) {
        editor.remove(key);
        editor.apply();
    }

    @Nullable
    public <T> T get(Class<T> clazz, String key) {
        try {
            String resValue = sp.getString(key, null);

            if (resValue == null) {
                return null;
            }

            if (clazz.isAssignableFrom(String.class)) {
                return (T) resValue;
            }

            if (clazz.isAssignableFrom(Integer.class) || clazz.isAssignableFrom(int.class)) {
                return (T) Integer.valueOf(resValue);
            }

            if (clazz.isAssignableFrom(Boolean.class) || clazz.isAssignableFrom(boolean.class)) {
                return (T) Boolean.valueOf(resValue);
            }

            if (clazz.isAssignableFrom(Long.class) || clazz.isAssignableFrom(long.class)) {
                return (T) Long.valueOf(resValue);
            }

            if (clazz.isAssignableFrom(Float.class) || clazz.isAssignableFrom(float.class)) {
                return (T) Float.valueOf(resValue);
            }

            if (clazz.isAssignableFrom(JSONObject.class)) {
                return (T) new JSONObject(resValue);
            }


        } catch (Exception ex) {
            ex.printStackTrace();
        }


        return null;
    }


    public synchronized void removeAll() {
        for (String key : sp.getAll().keySet()) {
            remove(key);
        }
    }

}

File: app/src/main/java/ir/roohi/farshid/reminderpro/Storage.kt
package ir.roohi.farshid.reminderpro

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences

import org.json.JSONObject


/**
 * Created by ImanX.
 * MyZarinPal-Android-V4 | Copyrights 2018 ZarinPal Crop.
 */

class Storage @SuppressLint("CommitPrefEdits")
constructor(context: Context, name: String) {

    private val sp: SharedPreferences = context.getSharedPreferences(name, Context.MODE_PRIVATE)
    private val editor: SharedPreferences.Editor

    constructor(context: Context) : this(context, "dev") {}

    init {
        this.editor = sp.edit()
    }

    fun <T> put(clazz: Class<T>, key: String, value: T) {
        editor.putString(key, value.toString())
        editor.apply()
    }


    private fun remove(key: String) {
        editor.remove(key)
        editor.apply()
    }

    operator fun <T> get(clazz: Class<T>, key: String): T? {
        try {
            val resValue = sp.getString(key, null) ?: return null

            if (clazz.isAssignableFrom(String::class.java)) {
                return resValue as T
            }

            if (clazz.isAssignableFrom(Int::class.java) || clazz.isAssignableFrom(Int::class.javaPrimitiveType!!)) {
                return Integer.valueOf(resValue) as T
            }

            if (clazz.isAssignableFrom(Boolean::class.java) || clazz.isAssignableFrom(Boolean::class.javaPrimitiveType!!)) {
                return java.lang.Boolean.valueOf(resValue) as T
            }

            if (clazz.isAssignableFrom(Long::class.java) || clazz.isAssignableFrom(Long::class.javaPrimitiveType!!)) {
                return java.lang.Long.valueOf(resValue) as T
            }

            if (clazz.isAssignableFrom(Float::class.java) || clazz.isAssignableFrom(Float::class.javaPrimitiveType!!)) {
                return java.lang.Float.valueOf(resValue) as T
            }

            if (clazz.isAssignableFrom(JSONObject::class.java)) {
                return JSONObject(resValue) as T
            }


        } catch (ex: Exception) {
            ex.printStackTrace()
        }


        return null
    }


    @Synchronized
    fun removeAll() {
        for (key in sp.all.keys) {
            remove(key)
        }
    }

}
